collect_app/src/main/java/org/odk/collect/android/activities/FormMapActivity.java

```diff
@@ -56,15 +64,45 @@
 
     private FormMapViewModel viewModel;
 
-    private MapFragment map;
+    @VisibleForTesting public MapFragment map;
+
+    /**
+     * Quick lookup of instance objects from map feature IDs.
+     */
+    private final Map<Integer, FormMapViewModel.MappableFormInstance> instancesByFeatureId = new HashMap<>();
+
+    /**
+     * Points to be mapped. Note: kept separately from {@link #instancesByFeatureId} so we can
+     * quickly zoom to bounding box.
+     */
+    private final List<MapPoint> points = new ArrayList<>();
+
+    /**
+     * True if the map viewport has been initialized, false otherwise.
+     */
+    private boolean viewportInitialized;
+
+    @VisibleForTesting public ViewModelProvider.Factory viewModelFactory;
 
     @Override public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
         FormsDao dao = new FormsDao();
-        Form form = dao.getFormsFromCursor(dao.getFormsCursor(getIntent().getData())).get(0);
-        viewModel = ViewModelProviders.of(this,
-                new FormMapActivity.FormMapViewModelFactory(form, new DatabaseInstancesRepository())).get(FormMapViewModel.class);
+        Form form = null;
+        List<Form> forms = dao.getFormsFromCursor(dao.getFormsCursor(getIntent().getData()));
+        if (forms.size() == 1) {
+            form = forms.get(0);
+        } else {
+            // This shouldn't be possible because the instance URI had to be used in the calling
+            // activity but it has been logged to Crashlytics.
+            finish();
+        }
+
+        if (viewModelFactory == null) { // tests set their factories directly
```

Comment:
Is it possible to do this at field initialization or do we need to be in `onCreate` for some things to work?

Reply:
The content provider URI of the form is passed via intent to the activity. That feels like way too low-level of a concern for `FormMapViewModel` to have which is why I resolved it to a `Form` in `onCreate`. If the `Form` is passed in to the factory constructor, then the factory can't be built at the field initialization. I suppose I could use a setter on the factory? Or do you think the view model should get the form uri?

Reply:
Nah I think I like the way you have it then. I haven't found a nice way to set these up in `onCreate`. In `FormEntryActivity` I've ended up with `setupViewModels` method to try and seperate it out from all the other `onCreate` logic.